### PR TITLE
Sparse Matrix Filter

### PR DESCRIPTION
**Pull request by johnc1231**
This allows filtering sparse matrices without densifying.

### CODE REVIEW

**Comment by johnc1231** [comment]
While this seems to currently be a functioning improvement on what we had before, it's still slower than I'd like it to be for Jacob's use case and getting unexpected OOM errors, working on it now.

File: hail/src/main/scala/is/hail/linalg/GridPartitioner.scala [hunk 1]
```diff
@@ -25,9 +31,9 @@ case class GridPartitioner(blockSize: Int, nRows: Long, nCols: Long, maybeBlocks
 
   val maxNBlocks: Long = nBlockRows.toLong * nBlockCols
   
-  require(maybeBlocks.forall(bis => bis.isEmpty ||
+  assert(maybeBlocks.forall(bis => bis.isEmpty ||
     (bis.isIncreasing && bis.head >= 0 && bis.last < maxNBlocks &&
-      bis.length < maxNBlocks))) // a block-sparse matrix cannot have all blocks present
+      bis.length < maxNBlocks)), maybeBlocks.map(_.toIndexedSeq)) // a block-sparse matrix cannot have all blocks present
```
[the remark this answers is not in the document]
Comment (johnc1231) [reply hail/src/main/scala/is/hail/linalg/GridPartitioner.scala]: Great, that's what I'll do.

**Comment by johnc1231** [comment]
Did you have additional comments on this?

**Comment by patrick-schultz** [comment]
No, I'm ready to approve. Are you done working on it?

**Comment by johnc1231** [comment]
I'm done with this PR. The type inference issue got addressed in another PR, and I'll try and make this use less memory in the future, but as it is currently this is an improvement.

**Pull request merged**